### PR TITLE
makre sure defining VFP table for THP processing

### PR DESCRIPTION
**Pull request by GitPaean**
When we have non-defaulted THP constraint, we have to make sure to provide a VFP table for it. 

When we refer to a VFP table, we make sure the table is valid/defined. 

Originally, the plan was to checking whether VFP table is properly defined from the simulator, but it turns out to be extremely difficult to do it well/properly. The main point was to trigger the throw before all the possible circumstances that can utilize it. 

At the end, it can be the decent solution to do it from the parser side, avoiding to consider all kinds of complicated SCHEDULE setup.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
The drawback of current approach is that if the THP constraint is a UDQ value (which can be zero), we consider that we have to define a VFP table for that also. 

Not totally sure whether we handle zero THP constraint resulting from UDQ in the simulator side properly either.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/input/eclipse/Schedule/Well/WellInjectionProperties.cpp [hunk 1]
```diff
@@ -106,14 +108,19 @@ namespace Opm {
         } else
             this->dropInjectionControl(InjectorCMode::RESV);
 
+        this->VFPTableNumber = record.getItem("VFP_TABLE").get< int >(0);
```
Comment (GitPaean) [inline opm/input/eclipse/Schedule/Well/WellInjectionProperties.cpp:111]: needs to check whether VFP table is available.

**Comment by GitPaean** [comment]
I am marking the PR as ready for review now.

File: opm/input/eclipse/Schedule/Well/Well.hpp [hunk 2]
```diff
@@ -177,7 +178,8 @@ class Well {
         void handleWCONINJE(const DeckRecord& record,
                             const double bhp_def,
                             bool availableForGroupControl,
-                            const std::string& well_name);
+                            const std::string& well_name,
+                            const KeywordLocation& location);
```
Comment (bska) [inline opm/input/eclipse/Schedule/Well/Well.hpp:182]: Would you mind including the new parameter in the Doxygen description too?

File: opm/input/eclipse/Schedule/Well/Well.hpp [hunk 3]
```diff
@@ -306,7 +308,8 @@ class Well {
                             const double bhp_def,
                             const UnitSystem& unit_system,
                             const std::string& well,
-                            const DeckRecord& record);
+                            const DeckRecord& record,
+                            const KeywordLocation& location);
```
Comment (bska) [inline opm/input/eclipse/Schedule/Well/Well.hpp:312]: Would you mind including the new parameter in the Doxygen description too?

**Review by bska — APPROVED** [review]
> The drawback of current approach is that if the THP constraint is a UDQ value (which can be zero), we consider that we have to define a VFP table for that also.

I don't think we need to worry about that.  If a model uses a user-defined quantity for its THP constraint, then I think we can assume that the UDQ will be non-zero for at least portions of the simulation run which would necessitate having lift curve for the well in question.  It's possible may need to do some additional checking on the simulator side for a zero THP constraint value when determining whether or not the constraint is violated, but I think we can defer that until the situation arises.

Anyway, this PR looks good to me apart from a couple of minor issues.

File: opm/input/eclipse/Schedule/Well/WellProductionProperties.cpp [hunk 4]
```diff
@@ -193,9 +196,20 @@ namespace Opm {
...
-                if (cmode.first == "THP" && this->THPTarget.is<double>() && this->THPTarget.zero())
-                    continue;
+                if (cmode.first == "THP") {
+                    // a zero value THP limit will not be handled as a THP limit
+                    if (this->THPTarget.is<double>() && this->THPTarget.zero()) {
+                        continue;
+                    } else {
+                        // make sure we specify a VFP table for it
+                        // const int vfp_table = record.getItem("VFP_TABLE").get<int>(0);
+                        if (this->VFPTableNumber == 0) {
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/input/eclipse/Schedule/Well/WellProductionProperties.cpp:206]: Sure.

**Comment by GitPaean** [comment]
Thanks for the prompt review and comments.

**Comment by GitPaean** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now and I'll merge into master.

**Pull request merged**